bypass initial prompt when calling soup

petiepooo opened the pull request:
By calling soup with the -y flag here, it will not prompt before rebooting; the system will just disconnect.  If the state of the system is not such that it will survive a reboot, you've just caused an administrative headache.

petiepooo committed:
remove -y flag when calling soup

Never.  Ever.  Ever.  Ever cause a system to reboot without explicit consent from the administrator!  Ever!  That is very unfriendly behavior.

By calling soup with the -y flag here, it will not prompt before rebooting; the system will just disconnect.  If the state of the system is not such that it will survive a reboot, you've just caused an administrative headache.

dougburks commented:
Hi @petiepooo ,

Thanks for the PR!

Before `so-elastic-features` calls `soup`, it should run the following:
```
header "Please wait while verifying your system is up to date..."
apt update
UPDATES_AVAILABLE=$(so-apt-check 2>&1)
if [ "$UPDATES_AVAILABLE" != "0;0" ]; then
        echo
        echo "Your system is missing some updates!"
        echo "Please run 'sudo soup' and follow any instructions including rebooting if necessary."
        echo "Then verify that everything is working correctly before trying to run this again."
        exit
fi
```

The intention here is to make sure that the system is fully updated and rebooted before we ever run `soup`.

Did that not work in your case?

petiepooo commented:
Hi Doug,
I avoid soup like the plague.  (Hrmm.. should that expression be updated for COVID??)
As a long-time sysadmin, I prefer to use the administrative tools Ubuntu provides natively for updating my sensors: apt for updating packages and docker pull for updating images.  Soup is made for non-administrators to have an easy way to keep things updated, and does things like restart services I've intentionally stopped, and restart elastic instances right before it prompts to reboot (if it even prompts).  It also doesn't account for the other apps and services I'm running on this box.  If you recall, one of the first patches I submitted was https://groups.google.com/u/1/g/security-onion/c/k9Sp_RiEHRo six years ago to update the securityonion-pfring-modules package so that there was no need to treat it like a special snowflake as soup did.
Since your reply indicates that you are resistant to this pull request here are some additional thoughts:
1) What does so-elastic-features even care if there are some packages in need of upgrade or a reboot pending?  I've long scrolled past the incident, but in this case, I think there may have been something pending removal, or maybe I had updated something like libc but hadn't yet rebooted as I surely would have soon.  In any case, I did not get the "missing updates" message, as it did not exit the script at that point.
2) So you understand, my typical upgrade cycle goes like this: apt --download-only any available packages so they are cached locally;  if there's a new so-elastic package, pre-pull the new images to go with it;  if there's a new kernel available, install it and autoremove any older kernels;  install/upgrade non-securityonion apps that don't interrupt service;  if there are securityonion-* or other packages pending that would interrupt service, run so-sensor-stop for log continuity;  if there's a new docker version, stop other docker containers and install it;  if there's a new zeek package, install it, stop zeek (again!!), and reapply my updates to its config;  if there's a new elastic package, install it but pkill the so-elastic-start script run by the pkg's postinst script from another shell (since I'm about to reboot anyway), and let it complete;  install any remaining updates available, stop other apps as needed, and finally reboot (sometimes I'll run soup here just to let it do the reboot).  Once back online, I remove any old docker images, and verify all services.  Doing it this way, most apps are already stopped when the reboot happens, service downtime is minimized, logs are coherent (eg no nids alerts without backing zeek logs), and only one reboot is needed.  I'm fully aware this is not the supported method, and I'm fully okay with that.
3) I assume so-elastic-features runs soup primarily to download the latest securityonionsolutionselas/so-* images.  Since it sounds like you don't care for this PR, would you be amenable to extracting just the portion of soup that deals with elastic into a separate script (such as so-elastic-upgrade-images)?  That script could be called at the same point in soup as what is pulled out.  That way, there's less dependency on the securityonion-sostat package within the securityonion-elastic package, and running so-elastic-features can call that more targeted script instead of the sledgehammer that is soup.  Reducing dependencies between modules is a good goal, I think...
4) Please don't tell me I should be running soup.  Soup has it's place, but I prefer fine control of the update process rather than relying on an error-prone script that doesn't account for my platform's personalizations.  While I can read scripts and perform their actions manually, I prefer to use pre-generated and supported scripts wherever they work for me.  Soup doesn't work for me, and so-elastic-features doesn't either as long as it can trigger an unannounced reboot.

dougburks commented:
Hi @petiepooo ,

> What does so-elastic-features even care if there are some packages in need of upgrade or a reboot pending?

When we decided to integrate with Elastic Features, we had to update several other packages to support that integration.  `so-elastic-features` therefore needs to ensure that those packages have been fully updated before switching over to the new containers.

The `-y` option on `soup` was used mainly to skip `soup`'s initial prompt, so perhaps a simple compromise here might be to change `soup -y` to `echo | soup`.

Thoughts?

petiepooo committed:
send a single CR to bypass first prompt in soup

This should bypass first prompt, but if a reboot is called for, still give the option to cancel and return to CLI.

petiepooo commented:
I think that's a great solution, but will need some testing.  It will clutter output a bit more showing the first prompt, but what will happen on a second prompt; will the broken pipe abort soup early or something else?  Should it be called as `(echo; cat) | soup` so it sends a CR first but then reverts to stdin?

I'm also not positive how the situation was originally triggered for me; it caught me by surprise and I didn't save the cli output.  I suspect /var/run/reboot-required was present for previous upgrades I had just run in (soup line 346).  Dropping that file should be enough to duplicate.

Thank you, Doug.

petiepooo committed:
fix call to soup so it actually skips first prompt

This echoes a single return first, then reads everything from stdin via cat.  Without cat, all reads just return since stdin is EOF.

petiepooo commented:
Looks like that cat is needed after all to switch to stdin...

> p@host:\~$ cat ./test
#!/bin/bash
echo prompt 1
read IN
echo prompt 2
read IN
echo prompt 3
p@host:\~$ echo | ./test
prompt 1
prompt 2
prompt 3
p@host:\~$ ( echo; cat ) | ./test
prompt 1
prompt 2
^C
p@host:\~$

dougburks commented:
Merged, thanks @petiepooo !

dougburks commented:
After testing this, I noticed a problem. If soup doesn't actually prompt the user for input (meaning that all updates have already been installed), then it ends up just hanging there waiting for the user to press Enter but with no indication to actually do so.

At this point, I think the quickest, easiest, and safest solution is to go back to your original proposal of just using `soup` (without the `-y` option).  It will mean that users have to press the Enter key to proceed, but I think that's the best path forward.
